Introducing support to Melnor RainCloud sprinkler systems

tchellomello opened the pull request:
## Description:
Introducing support to the Melnor RainCloud sprinkler system on Home Assistant. 

This PR adds some new `binary_sensor`, `switch` and `sensor` components.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3304


![image](https://user-images.githubusercontent.com/809840/30013380-3a69f0a0-9114-11e7-8a87-49910d141ee8.png)


## Example entry for `configuration.yaml` (if applicable):
```yaml
# configuration.yaml
raincloud:
  username: user@domain
  password: secret
  watering_time: 30

binary_sensor:
  - platform: raincloud

sensor:
  - platform: raincloud

switch:
  - platform: raincloud
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

tchellomello committed:
Introducing support to Melnor RainCloud sprinkler systems

tchellomello committed:
Make monitored_conditions optional for sub-components

pvizeli commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,95 @@
+        raincloud = RainCloudy(username=username, password=password)
+        if not raincloud.is_connected:
+            return False
+        hass.data['raincloud'] = RainCloudHub(hass,
```
use a constante for 'raincloud' like `DATA_`

pvizeli commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,95 @@
+
+        # Load data
+        track_time_interval(hass, self._update_hub, SCAN_INTERVAL_HUB)
+        self._update_hub(SCAN_INTERVAL_HUB)
```
move this into `async_added_to_hass`

pvizeli commented on homeassistant/components/sensor/raincloud.py:
```diff
@@ -0,0 +1,108 @@
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up a sensor for a raincloud device."""
+    raincloud = hass.data.get('raincloud').data
```
same

pvizeli commented on homeassistant/components/sensor/raincloud.py:
```diff
@@ -0,0 +1,108 @@
+                self._icon = 'mdi:battery'
+            else:
+                self._icon = 'mdi:battery-{}'.format(str(rounded_level))
+        return self._icon
```
Use the battery helper from `utils/icon.py`

pvizeli commented on homeassistant/components/sensor/raincloud.py:
```diff
@@ -0,0 +1,108 @@
+        attrs[ATTR_ATTRIBUTION] = CONF_ATTRIBUTION
+        attrs['identifier'] = self._data.serial
+        attrs['current_time'] = self._data.current_time
+        return attrs
```
same

pvizeli commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,95 @@
+class RainCloudHub(Entity):
+    """Base class for all Raincloud entities."""
+
+    def __init__(self, hass, data, default_watering_timer):
```
remove hass from init

tchellomello replied:
I had to re-add this due to the dispatcher signal.

pvizeli reviewed (changes requested):
I think you should update the data inside `setup` with a function and track_time_interval. After that fire a dispatcher event that update the entity like andorid_ip_webcam. That save a lot of resource. Add also a `scan_interval` option to you component

MartinHjelmare commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,95 @@
+    return True
+
+
+class RainCloudHub(Entity):
```
You shouldn't inherit from `Entity` if you won't inherit this class in the entities that you add to home assistant. Just make a data class that inherits from object.

pvizeli replied:
Good point

tchellomello commented:
Thanks guys.. I'll work on the modifications.  :+1:

tchellomello committed:
Part 1/2 - Modified attributes, added DATA_ constant and using battery helper

tchellomello committed:
Part 2/2 - Refactored self-update hub

MartinHjelmare commented:
I don't see where you connect a target function to the signal in the dispatcher.

pvizeli commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,97 @@
+        raincloud = hass.data[DATA_RAINCLOUD]
+        raincloud.data.update()
+
+        dispatcher_send(hass, SIGNAL_UPDATE_RAINCLOUD, raincloud)
```
You need not send the hub object, there is allready availaible.

pvizeli commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,97 @@
+        dispatcher_send(hass, SIGNAL_UPDATE_RAINCLOUD, raincloud)
+
+    # Call the Raincloud API to refresh updates
+    track_time_interval(hass, hub_refresh, SCAN_INTERVAL)
```
Use the config value for the interval instead the statit SCAN_INTERVAL

pvizeli reviewed (changes requested):
You need connect to the dispatcher singal like this: https://github.com/home-assistant/home-assistant/blob/90f9a6bc0ab05ce4f633a4309bde326b16a44bac/homeassistant/components/sensor/envisalink.py#L53-L59

tchellomello committed:
Fixed change requested:
- Dispatcher signal connection
- Don't send raincloud object via dispatcher_send()
- Honoring the dynamic scan_interval value on track_time_interval()

tchellomello commented:
@MartinHjelmare  @pvizeli  Thanks for your help reviewing the code guys.

MartinHjelmare commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,112 @@
+        self.default_watering_timer = default_watering_timer
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
```
You should put this in the entity device classes, or make an intermediary entity class that has this which you can inherit from in the device classes.

tchellomello committed:
Inherents async_added_to_hass() on all device classes

tchellomello committed:
Makes lint happy

MartinHjelmare commented on homeassistant/components/binary_sensor/raincloud.py:
```diff
@@ -0,0 +1,103 @@
+    return True
+
+
+class RainCloudBinarySensor(RainCloudHub, BinarySensorDevice):
```
Don't inherit from RainCloudHub. If you want to have `async_added_to_hass` in an intermediate class, make a separate class for that that inherits from `Entity`. Keep `RainCloudHub` as a pure data class separate from the entity.

MartinHjelmare commented on homeassistant/components/binary_sensor/raincloud.py:
```diff
@@ -0,0 +1,103 @@
+
+    def __init__(self, hass, data, sensor_type):
+        """Initialize a sensor for raincloud device."""
+        super().__init__(hass, data)
```
Remove this.

MartinHjelmare replied:
If you move the init to the intermediate class, you don't even need this line.

MartinHjelmare commented on homeassistant/components/binary_sensor/raincloud.py:
```diff
@@ -0,0 +1,103 @@
+    def __init__(self, hass, data, sensor_type):
+        """Initialize a sensor for raincloud device."""
+        super().__init__(hass, data)
+        self._hass = hass
```
Remove this.

MartinHjelmare commented on homeassistant/components/binary_sensor/raincloud.py:
```diff
@@ -0,0 +1,103 @@
+class RainCloudBinarySensor(RainCloudHub, BinarySensorDevice):
+    """A sensor implementation for raincloud device."""
+
+    def __init__(self, hass, data, sensor_type):
```
Don't pass `hass`, see above.

MartinHjelmare replied:
I suggest to move the whole init method to the intermediate class, since most device init method have the same logic. Only the switch class is different.

MartinHjelmare reviewed (changes requested):
Don't conflate concerns.

MartinHjelmare commented on homeassistant/components/switch/raincloud.py:
```diff
@@ -0,0 +1,119 @@
+
+    def __init__(self, hass, data, sensor_type, default_watering_timer):
+        """Initialize a switch for raincloud device."""
+        super().__init__(hass, data)
```
You then call super init here with all args except the watering timer.

```py
super().__init__(*args)
```

MartinHjelmare commented on homeassistant/components/switch/raincloud.py:
```diff
@@ -0,0 +1,119 @@
+class RainCloudSwitch(RainCloudHub, SwitchDevice):
+    """A switch implementation for raincloud device."""
+
+    def __init__(self, hass, data, sensor_type, default_watering_timer):
```
If you move the init to the intermediate class, you overwrite the init in this class since it has a special case with the watering timer.

```py
def __init__(self, default_watering_timer, *args):
```

MartinHjelmare commented on homeassistant/components/switch/raincloud.py:
```diff
@@ -0,0 +1,119 @@
+        self._hass = hass
+        self._sensor_type = sensor_type
+        self.data = data
+        self._default_watering_timer = default_watering_timer
```
And you just keep this line.

MartinHjelmare commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,107 @@
+
+    def _update_callback(self):
+        """Callback update method."""
+        self.data.update()
```
Isn't the point of the signal to update the entity state? The data has already been updated before the signal is dispatched at line 81 above.

tchellomello committed:
* Refactored RainCloud code to incorporate suggestions.
  Many thanks to @pvizelli and @martinhjelmare!!

pvizeli commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,181 @@
+
+    def _update_callback(self):
+        """Callback update method."""
+        self.update()
```
I think you should use `self.schedule_update_ha_state(True)`

tchellomello replied:
Thanks!

MartinHjelmare commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,181 @@
+    """Base class for all Raincloud entities."""
+
+    def __init__(self, data):
+        """Initialize the entity."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/raincloud.py:
```diff
@@ -0,0 +1,181 @@
+
+
+class RainCloudHub(object):
+    """Base class for all Raincloud entities."""
```
Stale docstring.

MartinHjelmare reviewed (changes requested):
Nice!

tchellomello commented on homeassistant/components/sensor/raincloud.py:
```diff
@@ -0,0 +1,70 @@
+    return True
+
+
+class RainCloudSensor(RainCloudEntity, Entity):
```
Yes, makes sense!!

tchellomello committed:
Removed Entity from RainCloud sensor and fixed docstrings

tchellomello commented:
@pvizeli  @MartinHjelmare  thank you so much for your assistance and guidance!! It is very appreciated!!

pvizeli committed:
Update raincloud.py

pvizeli committed:
Update raincloud.py

pvizeli committed:
fix lint

pvizeli commented:
Good teamwork @tchellomello @MartinHjelmare
